File: Dynamic_Programming_04_Abbrevition.cpp
// code_report Solution

// Solution 1:
// https://youtu.be/DW92IHf8KK8

#include <unordered_set>
#include <algorithm>
#include <string>

using namespace std;

void rec(string A, string B);

unordered_set<string> memo;
bool possible;

auto abbreviation(string a, string b) {
   possible = false;
   memo.clear();
   rec(a, b);
   return possible ? "YES" : "NO";
}

void rec(string A, string B) {
   if (possible || A.size() < B.size()) return;
   if (B.empty()) {
      if (all_of(A.begin(), A.end(), [](auto e) { return islower(e); }))
         possible = true;
      return;
   }

   auto p = memo.insert(A + '#' + B);
   if (!p.second) return; // return if A#B already in set

   auto fc = A[0]; // first char 
   A.erase(0, 1);
   if (islower(fc)) rec(A, B);
   if (toupper(fc) != B[0]) return;
   B.erase(0, 1);
   rec(A, B);

   return;
}

// Solution 2:
// https://youtu.be/ViILdd8495M

auto abbreviation(string a, string b) {
   int n = a.size();
   int m = b.size();
   vector<vector<int>> dp(n + 1, vector<int>(m + 1));
   dp[0][0] = 1;

   for (int i = 0; i < n; i++) {
      for (int j = 0; j <= m; j++) {
         if (!dp[i][j]) continue;
         if (j < m && toupper(a[i]) == b[j]) dp[i + 1][j + 1] = 1;
         if (!isupper(a[i]))                 dp[i + 1][j    ] = 1;
      }
   }

   return (dp[n][m] ? "YES\n" : "NO\n");
}

File: Dynamic_Programming_04_Abbrevition.py
# code_report Solution

# Solution 1:
# https://youtu.be/DW92IHf8KK8

memo = set()
possible = False

def abbreviation(a, b):
    global possible, memo
    possible = False
    memo.clear()
    rec(a,b)
    return "YES" if possible else "NO"

def rec(a, b):
    global possible, memo
    if (possible or len(a) < len(b)):
        return
    if (len(b) == 0):
        if (a.islower() or len(a) == 0):
            possible = True
        return
    
    t = a+'#'+b
    if (t in memo):
        return
    memo.add(t);
    
    fc = a[0:1] # first char
    a = a[1:]
    if (fc.islower()):
        rec(a,b)
    if (fc.upper() != b[0:1]):
        return
    b = b[1:]
    rec(a,b)
    
    return

# Solution 2:
# https://youtu.be/ViILdd8495M

def abbreviation(a, b):
    n, m = len(a), len(b)
    dp = [[0 for i in range(m+1)] for j in range(n+1)]
    dp[0][0] = 1
    
    for i in range(n):
        for j in range(m+1):
            if not dp[i][j]: continue;
            if j < m and a[i].upper() == b[j]: dp[i+1][j+1] = 1;
            if not a[i].isupper():             dp[i+1][j  ] = 1;
    
    return 'YES' if dp[n][m] else 'NO'
